cohttp-lwt/src/client.ml

```diff
@@ -1,7 +1,9 @@
 open Lwt.Infix
 module Header = Cohttp.Header
 
-module Make (IO : S.IO) (Net : S.Net with module IO = IO) = struct
+exception Retry
+
+module Make (IO : S.IO) (Net : S.Net with module IO = IO) (Time : Mirage_time.S) = struct
```

Comment:
Instead of requiring the Mirage_time signature, what about specifying only the part that we need? This will make it both more flexible to users and make allow us to drop the mirage dependency.

Reply:
`Mirage_time` is
```
(** Sleep operations. *)
module type S = sig
  val sleep_ns: int64 -> unit Lwt.t
  (** [sleep_ns n] Block the current thread for [n] nanoseconds, treating
      the [n] unsigned.  *)
end
```
There are no other parts to pick from. `mirage-time-unix` depends on `duration`. This dependency could easily be dropped. What do you propose?

Reply:
If the signature is so simple then I definitely suggest inlining it.

Reply:
Note that this choice becomes incompatible with `functoria` which passes implementation _via_ _functors_. Not sure that we want to break that. We can pass that by a value `~time` (as I did into `dream`) but at the end, a _functor_ is required.

Reply:
uh, let me push what I did…

Reply:
What about 4efebd7 ?
I inlined the signature of `Mirage_time.S` in `Cohttp_lwt` and called it `Sleep`, but kept the dependency on `mirage-time` in `Cohttp_mirage`:
```
module Cohttp_mirage.Client.Make : sig
    (P : Mirage_clock.PCLOCK)
    (T : Mirage_time.S)
    (R : Resolver_mirage.S)
    (S : Conduit_mirage.S) : sig
```
while in unix I do
```
module Sleep = struct
  let sleep_ns ns =
    Lwt_unix.sleep @@
    let open Int64 in
    if ns >= zero
    then to_float ns
    else (shift_right_logical ns 1 |> to_float) *. 2.
end

include Cohttp_lwt.Make_client (Io) (Net) (Sleep)
```

Is this ok with `functoria` @dinosaure ?
Is it ok to keep the dependency of `cohttp-mirage` on `mirage-time` @rgrinberg ?

Reply:
See this [comment](https://github.com/mirage/ocaml-cohttp/pull/853#discussion_r800618552). Again, I'm not really against that and it can be fine for `cohttp-lwt` but, for `cohttp-mirage`, it must be a `functor` (and it should be `Mirage_time.S`). This is what I did about `dream` which expects a `~time` argument to launch the server and be free about the operating system. Then, `dream-mirage` ensure that it expects a `Mirage_time.S` and we pass the function `Time.sleep_ns` to the runner then.

Reply:
Thanks for looking into it

At least `cohttp-mirage` will depend on (and use) `mirage-time`, so we have that package as a check for things to remain compatible.

The current code is already functorized, so that should be fine unless I misunderstood something

Reply:
So keep it like it like it is now or may I remove `Cohttp_lwt.S.Sleep` again and use `Mirage_time.S` in both `cohttp-lwt` and `cohttp-mirage` instead ? I would not depend on `mirage-time-unix`, but keep my inlined implementation of it.

Reply:
I don't have a strong opinion. Let's wait for @rgrinberg

Reply:
I insist on dropping the dependency. If our goal is to maintain compatibility with some signature coming from a mirage package, we can do it in the cohttp-mirage package.

Reply:
At a higher level, I don't quite understand why the Client module needs to add a Time dependency at all. Can't the connection cache be built _using_ the Client module, and adding in timeouts and connection pooling there?  This way existing Client consumers aren't affected.